### PR TITLE
ssl tests do not pass out of the box for openjdk 1.8.0_212.  test cipher no longer accepted.

### PR DESCRIPTION
**Pull request by fcnorman**
keytool reported that keystore.p12 was empty.

Added procedures and documentation to make ssl tests pass for Travis as well as remote contributors.

And, added findAvailableTcpPort for Ssl tests instead of assuming 8443 is free.

### CODE REVIEW

**Comment by fcnorman** [comment]
Travis PR checks are failing on what appears to be an erroneous checkstyle error in a module I did not touch.  It thinks GenericUtils.java has an unused import, ClassUtils.  Even though ClassUtils is used in the class.  I will stop tinkering with this.  If you would like to incorporate the pull request, please Resolve the conflict in GenericUtils.

File: inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java [hunk 1]
```diff
@@ -164,7 +164,7 @@ protected TypeMirror interfaceGenericTypeFor(TypeElement element, Class interfac
      * @param interfaceName The interface
      * @return The generic type or null
      */
-    protected TypeMirror interfaceGenericTypeFor(TypeElement element, String interfaceName) {
+    private TypeMirror interfaceGenericTypeFor(TypeElement element, String interfaceName) {
```
Comment (jameskleeh) [inline inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java]: Why was this class changed?

**Comment by jameskleeh** [comment]
Merge master into your branch

File: inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java (hunk 1 above)
Comment (fcnorman) [reply inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java]: That was an IntelliJ suggestion while I was investigating the checkstyle issue.  You are right, it doesn't need to be changed.  I'll change it back and resubmit.

**Comment by fcnorman** [comment]
retest this please

**Comment by fcnorman** [comment]
Ok, James, this should be ready to go.  I reverted GenericUtils and ArgumentAnnotationMetadataSpec to the latest master, and they are passing tests now.  Occasionally, GenericUtils compilation fails on the ClassUtils import, which is some sort of odd intermittent issue.  But, the latest tests worked.

Please let me know if you have any other requested changes.

Kind regards,
   Frank

File: inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java [hunk 2]
```diff
@@ -181,7 +180,7 @@ protected TypeMirror interfaceGenericTypeFor(TypeElement element, String interfa
             DeclaredType declaredType = (DeclaredType) tm;
             Element declaredElement = declaredType.asElement();
             if (declaredElement instanceof TypeElement) {
-               TypeElement te = (TypeElement) declaredElement;
+                TypeElement te = (TypeElement) declaredElement;
```
Comment (jameskleeh) [inline inject-java/src/main/java/io/micronaut/annotation/processing/GenericUtils.java]: Please revert the changes to this file

File: inject-java/src/test/groovy/io/micronaut/inject/annotation/ArgumentAnnotationMetadataSpec.groovy [hunk 3]
```diff
@@ -95,4 +95,4 @@ interface TestApi {
         metadata.getValue(Named).get() == "foo"
     }
 
-}
+}
```
Comment (jameskleeh) [inline inject-java/src/test/groovy/io/micronaut/inject/annotation/ArgumentAnnotationMetadataSpec.groovy]: Revert this whitespace change

**Comment by fcnorman** [comment]
Ok, James.  Should be ready to go again.

Kind regards,
   Frank

**Comment by fcnorman** [comment]
I pulled master into my branch again today because GenericUtils - which I didn't touch - was showing a delta again.  The merge got rid of that delta, and it looks like everything is good to go again.  Please let me know if it isn't.

**Comment by jameskleeh** [comment]
So am I understanding correctly that in order for the tests to pass users will have to execute a shell script first?

**Comment by jameskleeh** [comment]
@fcnorman I pushed a branch that has the keystore.p12 generated using the script you have here https://travis-ci.org/micronaut-projects/micronaut-core/builds/536494542

The build passed which means it works for older JDKs and it passed locally for me using 212. Do you see any problems with this approach? I don't want to add steps to run the tests locally and adding complexity to the build should be avoided.

**Comment by fcnorman** [comment]
Yes.  People who want to run ./gradlew check and pass tests will need to
run the generateKeystore.sh script once.

Btw, I added keystore.p12 to the .gitignore, but it was previously
committed.

**Comment by fcnorman** [comment]
So, regarding your last two comments.  I agree we don't want to add complexity.  I went back to the original problem I was trying to solve, which is:  it would be great to be able to git clone git@github.com:micronaut-projects/micronaut-core, cd micronaut-core, ./gradlew check and get passing tests.  I just went to a brand new Ubuntu 19.04 environment and started all over.  I had to:  1) downshift java from jdk 11 to jdk 8, 2) install docker and fix docker permissions (otherwise 5 docker tests failed), and that left one SSL test failing (SslStaticCertSpec failing because of SSLHandshakeException:  No appropriate protocol...)  The exact Java I installed was openjdk 1.8.0_212 build 1.8.0_212-8u212-b03-0ubuntu1.19.04.2-b03.  Then, if I run the ./generateKeystore.sh in this PR, all tests pass.

One thing we could consider is:  in addition to adding the ./generateKeystore.sh invocation in the travis-build-pr.sh, should we also add it to the build.gradle so that it happens automatically when ./gradlew check is performed?  On the one hand it adds complexity to the build, but on the other hand it automates everything for the developer.

Please consider and let me know what I should do next.

**Comment by jameskleeh** [comment]
Can you verify the last step is no longer necessary with the branch I
linked above? The test should pass as is with that JDK and OS

**Comment by fcnorman** [comment]
Just pulled a fresh clone of micronaut-core, checked out the new-keystore branch and did a ./gradlew check.

ServerRedirectSpec test failed with ReadTimeoutException at youtubeClient.test().blockingGet()

**Comment by jameskleeh** [comment]
@fcnorman I really appreciate the effort put into this PR, however I don't see the need to complicate the build with the keystore creation given that the tests pass with the new keystore in the new-keystore branch on both older and the current JDK. I've made manually the changes to run the tests on a random port and updated the contributing documentation to match what you have here. I'll go ahead and merge the new-keystore branch and close this PR.

**Comment by fcnorman** [comment]
Sounds good!
